adapters/sspbc/sspbc.go

```diff
@@ -0,0 +1,354 @@
+
+			/*
+			  Determine bid type
+			  At this moment we only check if bid contains Adm property
+
+			  Later updates will check for video & native data
+			*/
+			if bid.AdM != "" {
+				bidType = openrtb_ext.BidTypeBanner
+			}
+
+			if BidExt, ok := a.adSlots[bidId]; ok {
+				var BidIdStored = BidExt.PbSlot
+				bid.ImpID = BidIdStored
+			}
+
+			// read additional data from proxy
+			var BidDataExt responseExt
+			if err := json.Unmarshal(bid.Ext, &BidDataExt); err != nil {
+				errors = append(errors, err)
+			} else {
+				var adCreationError error
+
+				/*
+					use correct ad creation method for a detected bid type
+					right now, we are only creating banner ads
+					if type is not detected / supported, throw error
+				*/
+				if bidType == openrtb_ext.BidTypeBanner {
+					bid.AdM, adCreationError = a.createBannerAd(bid, BidDataExt, internalRequest, seatBid.Seat)
+				} else {
+					adCreationError = fmt.Errorf("bid type is not supported: %s", bidType)
+				}
+
+				if adCreationError != nil {
+					errors = append(errors, err)
+				} else {
+					// append bid to responses
+					bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+						Bid:     &bid,
+						BidType: bidType,
+					})
+				}
+			}
+		}
+	}
+
+	return bidResponse, errors
+}
+
+func (a *adapter) createBannerAd(bid openrtb2.Bid, ext responseExt, request *openrtb2.BidRequest, seat string) (string, error) {
+	var mcad mcAd
+
+	if strings.Contains(bid.AdM, "<!--preformatted-->") {
+		// Banner ad is already formatted
+		return bid.AdM, nil
+	}
+
+	// create McAd payload
+	mcad.Id = request.ID
+	mcad.Seat = seat
+	mcad.SeatBid = make([]openrtb2.SeatBid, 1)
+	mcad.SeatBid[0].Bid = make([]openrtb2.Bid, 1)
+	mcad.SeatBid[0].Bid[0] = bid
+	mcMarshalled, err := json.Marshal(mcad)
+	if err != nil {
+		return bid.AdM, err
+	}
+
+	mcEncoded := base64.URLEncoding.EncodeToString(mcMarshalled)
+
+	bannerData := &templatePayload{
+		SiteId:  ext.SiteId,
+		SlotId:  ext.SlotId,
+		AdLabel: ext.AdLabel,
+		PubId:   ext.PublisherId,
+		Page:    request.Site.Page,
+		Referer: request.Site.Ref,
+		McAd:    mcEncoded,
+	}
+
+	// Prepare banner html, using template file
+	var filledTemplate bytes.Buffer
+	if err := a.bannerTemplate.Execute(&filledTemplate, bannerData); err != nil {
+		return bid.AdM, err
+	}
+
+	return filledTemplate.String(), nil
+}
+
+func getImpSize(Imp openrtb2.Imp) string {
+	impSize := "1x1"
+
+	if Imp.Banner != nil {
+		areaMax := int64(0)
+		for _, sizeI := range Imp.Banner.Format {
+			areaI := sizeI.W * sizeI.H
+			if areaI > areaMax {
+				areaMax = areaI
+				impSize = fmt.Sprintf("%dx%d", sizeI.W, sizeI.H)
+			}
+		}
+	}
+
+	// default fallback
+	return impSize
+}
+
+func (a *adapter) fillAdSlotData(impI openrtb2.Imp, impSize string) adSlotData {
+    var extData adSlotData
+    a.mux.Lock()
+    defer a.mux.Unlock()
+    // save slot data
+    a.adSizes[impSize] = a.adSizes[impSize] + 1
+    if a.adSlots[impI.ID].PbSlot != "" {
```

Comment:
This structure should help to detect placement on the page, when placement data (site id, placement id) is not provided. 

We do the same on prebid.js side, and if publisher does not supply id's, we will, for instance, send info that this adunit name = X, and  is third time that 300x250 placement was called. The second part (represented by above struct) will help if adUnit name is pseudorandom (as is the case with some publishers).

Buy You're right - while in js adapter case we have one instance of adapter running in browser environment, this will not work if there are multiple PBS instances running in a cluster.

I'm now removing sizeXuse mapping, and in the case of psudorandom placements we will resort to additional first party data supplied by the publisher.

Reply:
Done - adSizes are now longer kept in the adapter. AdSlots still are - we need that data between request and response - but that has been simplified.